torchvision/transforms/_geometry.py

```diff
@@ -0,0 +1,59 @@
+from typing import Any, Dict, Optional, Tuple
+
+import torch
+
+from torchvision.features import BoundingBox, Image
+
+from ._core import Transform, query_sample
+
+__all__ = ["HorizontalFlip", "RandomRotate", "RandomErase"]
+
+
+class HorizontalFlip(Transform):
+    @staticmethod
+    def image(input: Image) -> Image:
+        return Image(input.flip((-1,)))
+
+    @staticmethod
+    def bounding_box(input: BoundingBox) -> BoundingBox:
```

Comment:
Agreed. This way, when possible, we share the same functional implementation for images and bboxes and when not possible we implement new. Also we leave it up to the user to make the call depending on the use-case rather than assuming the BBox is in some part of the target.

See also this corner case transform that requires joint transform of all image, label, bbox:
https://github.com/pytorch/vision/blob/96f6e0a117d5c56f7e0237851dbb96144ebb110b/references/detection/transforms.py#L54-L129

Reply:
Not sure I understand:

> we share the same functional implementation for images and bboxes and when not possible we implement new

In most cases the implementation will be different for images, bounding boxes and possible other types. In any case, we will always have a separate method for each type. In case the transformation is the same, it is the developers responsibility to implement it with minimal duplications:

```python
class FooTransform(Transform):
    @staticmethod
    def _foo(input):
        return input

    @staticmethod
    def image(input: Image) -> Image:
        return FooTransform._foo(input)

    @staticmethod
    def bounding_box(input: BoundingBox) -> BoundingBox:
        return FooTransform._foo(input)
```

> Also we leave it up to the user to make the call depending on the use-case rather than assuming the BBox is in some part of the target.

That is actually what we are trying to avoid. If you want to, you can call `HorizontalFlip.image(...)`, but during normal operation you would do

```python
transform = HorizontalFlip()
transformed_image = transform(image)
transformed_bbox = transform(bbox)
```

This is needed to be able to transform a complete sample drawn from a dataset, which might include more than just an image. 

For a better overview of what we are trying to achieve, please read the [accompanying document](https://github.com/pmeier/torchvision-datasets-rework/blob/rfc/transforms/transforms.md) which is updated as this RFC progresses.

Reply:
@pmeier I didn't disagree with you. I just said I like the approach :)

What I meant is that there are a few transforms that can be implemented for images and bboxes in the same way. That can live in the functional API and as you said, we can call it for both `image()` and `bounding_box()`. But also the proposed API supports cases when this is not the case by providing separate implementations for the two.

> That is actually what we are trying to avoid. 

Yes I had a discussion with @fmassa who explained that you want to avoid this to make things composable. Sounds good. Just make sure you look the example I sent you because there are transforms where image, bbox and labels must be processed together. I think the API can cover for that but it's worth having this use-case in mind.